crates/bevy_ecs/src/query/access.rs

```diff
@@ -761,38 +761,61 @@ impl<T: SparseSetIndex> Access<T> {
         self.archetypal.ones().map(T::get_sparse_set_index)
     }
 
-    /// Returns an iterator over the component IDs that this `Access` either
-    /// reads and writes or can't read or write.
+    /// Returns an iterator over the component IDs and their [`ComponentAccessKind`].
     ///
-    /// The returned flag specifies whether the list consists of the components
-    /// that the access *can* read or write (false) or whether the list consists
-    /// of the components that the access *can't* read or write (true).
-    ///
-    /// Because this method depends on internal implementation details of
-    /// `Access`, it's not recommended. Prefer to manage your own lists of
-    /// accessible components if your application needs to do that.
-    #[doc(hidden)]
-    // TODO: this should be deprecated and removed, see https://github.com/bevyengine/bevy/issues/16339
-    pub fn component_reads_and_writes(&self) -> (impl Iterator<Item = T> + '_, bool) {
-        (
-            self.component_read_and_writes
-                .ones()
-                .map(T::get_sparse_set_index),
-            self.component_read_and_writes_inverted,
-        )
+    /// Returns `Err(UnboundedAccess)` if the access is unbounded.
+    /// This typically occurs when an [`Access`] is marked as accessing all
+    /// components, and then adding exceptions.
+    pub fn try_iter_component_access(
+        &self,
+    ) -> Result<impl Iterator<Item = ComponentAccessKind<T>> + '_, UnboundedAccess> {
+        if self.component_writes_inverted || self.component_read_and_writes_inverted {
+            return Err(UnboundedAccess);
```

Comment:
should we return component_writes_inverted and component_read_and_writes_inverted as part of the error? That way if the user has the archetype info they can figure out what components are available.